analytics/transform.go

```diff
@@ -197,16 +195,101 @@ func (event ParsedEvent) GetValue(field string) (interface{}, error) {
 	if err != nil {
 		return nil, err
 	}
-	if field == "contexts" || field == "derived_contexts" || field == "unstruct_event" {
-		// TODO: DRY HERE TOO?
-		output := make(map[string]interface{})
-		for _, pair := range kvPairs {
-			output[pair.Key] = pair.Value
-		}
-		return output, nil
+
+	return kvPairs, nil
+}
+
+// GetValue returns the value for a provided atomic field, without processing the rest of the event.
+// For unstruct_event, it returns a map of only the data for the unstruct event.
+func (event ParsedEvent) GetValue(field string) (interface{}, error) {
+	kvPairs, err := event.getParsedValue(field)
+	if err != nil {
+		return nil, err
 	}
 	return kvPairs[0].Value, nil
+}
+
+// GetUnstructEventValue returns the value for a provided atomic field inside an event's unstruct_event field
+//
+func (event ParsedEvent) GetUnstructEventValue(field string, path ...interface{}) (interface{}, error) {
+	if field != "unstruct_event" {
+		return nil, errors.New(`Invalid path, not inside unstruct_event`)
+	}
+
+	kvPairs, err := event.getParsedValue(field)
+	if err != nil {
+		return nil, err
+	}
+
+	// extract the key/value pairs of the event field into a map
+	eventMap := make(map[string]interface{})
+	for _, pair := range kvPairs {
+		eventMap[pair.Key] = pair.Value
+	}
 
+	j, err := json.Marshal(eventMap)
+	if err != nil {
+		return nil, err
+	}
+
+	b := make([]interface{}, len(path))
+	for idx := range path {
+		b[idx] = path[idx]
+	}
+
+	el := json.Get(j, b...)
+	return el.GetInterface(), nil
+}
+
+// GetContextValue returns the value for a provided atomic field inside an event's contexts or derived_contexts
+func (event ParsedEvent) GetContextValue(field string, path ...interface{}) (interface{}, error) {
+	if field != "contexts" && field != "derived_contexts" {
+		return nil, errors.New(`Invalid path, not a context`)
+	}
```

Comment:
This makes more sense here, however I wonder if this is what people expect? 

If we're being true to the event format, then the difference between `contexts` and `derived_contexts` is important (the latter are contexts added by enrichment, rather than trackers). However, in reality, would a user actually care if we think about where they are using this SDK, after the enriched stream?

I have a feeling if we could hide this implementation detail on the API to this function, then that might lead to a better experience for the users. Although it'll be a little messier inside this function. We perhaps have two options: 
1. Search `contexts` first, then `derived_contexts`, merge the results and return them.
2. Merge contexts and derived_contexts then find the searched for context.

Either way, worth bearing in mind that a "search" context could appear in either, neither or both of `contexts` and `derived_contexts` and it could appear multiple times across them.

Reply:
I am not sure I follow what you mean by merging the results.

Let's say both contexts have the `contexts_nl_basjes_yauaa_context_1` key inside them. The logical step would be to return a map, with a key/value pair for each of `contexts` and `derived_contexts`. 

Is this what you meant?

Reply:
So, all contexts must be considered as an array. It's completely valid to have two or more `contexts_nl_basjes_yauaa_context_1` attached to an event (they could be all in `contexts`, or one in `contexts` and others in `derived_contexts`, etc). It's rare that would be the case for many contexts (like `contexts_nl_basjes_yauaa_context_1` would rarely be attached twice) but imagine you had attached a `product` entity to a product listing page view, you might attach many product entities to the page view event to desribe all the products on the current page.

Notice here: https://github.com/snowplow/snowplow-golang-analytics-sdk/blob/3e5b416e1d1f5c783c7ea8194a039a8b59627ba5/analytics/shred.go#L120-L121
We are appending additional contexts to the existing key, if it already exists.

You could make this always return an array of the context data requested, e.g. 
```go
-> [ map["agentClass":"Special"], map["agentClass":"Special"] ]
```

Then users could select [0] if they just wanted the first one, or iterate through it if looking for a specific one.

Reply:
Done, output now looks like this:
``GetContextValue(`agentClass`)``

```go
[map[agentClass:Special] map[agentClass:Special] map[agentClass:Special]]
```

Reply:
I got it now, thank you.

I've added a contextName parameter to the function input and a check for the correct context in the value extraction.

Reply:
Also added the ability for the user to get the whole context by passing no field string.